src/pgwire/src/protocol.rs

```diff
@@ -1777,12 +1776,15 @@ fn describe_rows(stmt_desc: &StatementDesc, formats: &[mz_pgrepr::Format]) -> Ba
 }
 
 fn parse_sql(sql: &str) -> Result<Vec<Statement<Raw>>, ErrorResponse> {
-    mz_sql::parse::parse(sql).map_err(|e| {
-        // Convert our 0-based byte position to pgwire's 1-based character
-        // position.
-        let pos = sql[..e.pos].chars().count() + 1;
-        ErrorResponse::error(SqlState::SYNTAX_ERROR, e.message).with_position(pos)
-    })
+    match mz_sql::parse::parse_with_limit(sql) {
+        Ok(result) => result.map_err(|e| {
+            // Convert our 0-based byte position to pgwire's 1-based character
+            // position.
+            let pos = sql[..e.pos].chars().count() + 1;
+            ErrorResponse::error(SqlState::SYNTAX_ERROR, e.message).with_position(pos)
+        }),
+        Err(e) => Err(ErrorResponse::error(SqlState::PROGRAM_LIMIT_EXCEEDED, e)),
```

Comment:
This comes from Postgres

https://stackoverflow.com/questions/4936731/maximum-length-of-an-sql-query
https://github.com/postgres/postgres/blob/ef7002dbe0b06e4e5b42c89becd4eb9be2e9aa89/src/common/stringinfo.c#L300-L313